src/slice-chop.c

```diff
@@ -380,6 +380,244 @@ static SEXP chop_fallback_shaped(SEXP x, SEXP indices, struct vctrs_chop_info in
+  struct name_repair_opts name_repair_opts = new_name_repair_opts(name_repair, false);
+  PROTECT_NAME_REPAIR_OPTS(&name_repair_opts);
+
+  SEXP out = vec_unchop(x, indices, ptype, &name_repair_opts);
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP vec_unchop_indices(SEXP out,
+                               SEXP* p_out_names,
+                               PROTECT_INDEX out_pi,
+                               SEXP x,
+                               R_len_t x_size,
+                               SEXP indices,
+                               bool is_shaped,
+                               bool has_inner_names);
+
+static SEXP vec_unchop_sequentially(SEXP out,
+                                    SEXP* p_out_names,
+                                    PROTECT_INDEX out_pi,
+                                    SEXP x,
+                                    R_len_t x_size,
+                                    SEXP x_sizes,
+                                    bool is_shaped,
+                                    bool has_inner_names);
+
+static SEXP vec_unchop(SEXP x, SEXP indices, SEXP ptype, const struct name_repair_opts* name_repair) {
+  if (!vec_is_list(x)) {
+    Rf_errorcall(R_NilValue, "`x` must be a list");
+  }
+
+  R_len_t x_size = vec_size(x);
+
+  bool null_indices = (indices == R_NilValue);
+
+  // Apply size/type checking to `indices` before possibly exiting early from
+  // having a `NULL` common type
+  if (!null_indices) {
+    if (x_size != vec_size(indices)) {
+      Rf_errorcall(R_NilValue, "`x` and `indices` must be lists of the same size");
+    }
+    if (!vec_is_list(indices)) {
+      Rf_errorcall(R_NilValue, "`indices` must be a list of integers, or `NULL`");
+    }
+  }
+
+  ptype = PROTECT(vctrs_type_common_impl(x, ptype));
+
+  if (ptype == R_NilValue) {
+    UNPROTECT(1);
+    return R_NilValue;
+  }
+
+  x = PROTECT(vec_cast_common(x, ptype));
+
+  R_len_t out_size = 0;
+
+  SEXP x_sizes = vctrs_shared_empty_int;
+  if (null_indices) {
+    x_sizes = Rf_allocVector(INTSXP, x_size);
+  }
+  PROTECT(x_sizes);
+  int* p_x_sizes = INTEGER(x_sizes);
+
+  // `out_size` is computed from `indices` unless it is `NULL`
+  if (null_indices) {
+    for (R_len_t i = 0; i < x_size; ++i) {
+      R_len_t x_elt_size = vec_size(VECTOR_ELT(x, i));
+
+      out_size += x_elt_size;
+      p_x_sizes[i] = x_elt_size;
+    }
+  } else {
+    for (R_len_t i = 0; i < x_size; ++i) {
+      SEXP elt = VECTOR_ELT(x, i);
+
+      if (elt == R_NilValue) {
+        continue;
+      }
+
+      R_len_t index_size = vec_size(VECTOR_ELT(indices, i));
+      out_size += index_size;
+
+      // Each element of `x` is recycled to its corresponding index's size
+      elt = vec_recycle(elt, index_size, args_empty);
+      SET_VECTOR_ELT(x, i, elt);
+    }
+  }
+
+  indices = PROTECT(vec_as_indices(indices, out_size, R_NilValue));
+
+  PROTECT_INDEX out_pi;
+  SEXP out = vec_init(ptype, out_size);
```

Comment:
This could be `init(proxy(out))` rather than `proxy(init(out))` right? This seems more efficient in the general case since proxy might generate new data.

Reply:
We should do this in `vec_c()` too